src/include/simeng/pipeline/FetchUnit.hh

```diff
@@ -68,6 +75,19 @@ class FetchUnit {
   /** Clear the loop buffer. */
   void flushLoopBuffer();
 
+  /** Temporarily pause the FetchUnit. */
```

Comment:
We did talk over this option, but from initial trials it seems more complex to set up, and is likely to lead to a more inefficient context switching logic (all pipeline units will need flushing manually, compared to allowing the core to continue ticking until it is "empty enough" to halt safely.)

Reply:
Beyond the extra flushing being performed already, what else is there in terms of more manual flushing (ignoring the logic which is already implemented to flush the pipeline)?

Reply:
Each pipeline unit needs flushing individually. I think currently the only ones which have the capability (previously or added in this PR) are EU, DispatchIssue and ROB(?). I will take another attempt at converting to this model of context switch

Reply:
If you were to just flush the pipeline as per usual, wouldn't you just need to reset the RAT and then could safely context switch? I'm sure I'm forgetting some other structures but I think the idea still roughly holds. Are you concerned with the performance overhead of flushing the pipeline when compared to ticking the pipeline for an extra n cycles to finish the lifecycle of currently fetched instructions? I guess there may be something to be said regarding having to do duplicate work on individual non-speculated instructions that were in the pipeline prior to context switch...

Reply:
I guess not having to duplicate work for n instructions should outweigh a boolean check of `paused_` in fetch, setting `nextPC_` in the ROB, and in setting `nextPC_` Writeback (the last one being the most annoying).

Reply:
Adding to this. Even if we moved away from the current implementation of pausing the fetch unit I can't think of a way to track the last committed instruction's address in a different way to how `nextPC_` is stored; other than using an if statement and some kind of status flag. But at that point it would seem to add extra logic still to the ROB / WriteBack unit's commit logic...

In summary, in both implementation choices I think you'll need a boolean check (in fetch Unit or ROB), plus saving of `nextPC` in ROB / writeback unit. But, with the current choice you wouldn't have to re-execute any instructions.

Let me know if you have thoughts otherwise on how to make this more efficient

Reply:
I think finishing the current pipeline is a good optimisation regarding not having to do duplicate work so I vote for the current implementation. Regarding the setting to `nextPC_`, could we move it to fetch? When the unit is paused, it can record the current PC which will equate to the PC derived at retirement assuming no speculation goes wrong in the remaining pipeline and it isn't a branch. 

In the case there is an incorrect speculation, we could overwrite the PC value with that given in `updatePC`. The case where the final fetched instruction before the pause is a branch is a little trickier. We could disallow the final fetched instruction to be a branch? I can't think of any other solution that doesn't require extra logic in WB and ROB like we have now for detecting if the final instruction is a branch.

If we solve the branch issue, then we can remove the `nextPC_` usage in both WB and ROB, unless I've forgotten something here. Thoughts?

Reply:
Moving it to fetchUnit would be ideal as then for OoO we aren't having repeated copies of `nextPC_` stored. But I do wonder if the extra logic entailed for branch instructions would be less performant than storing the same value twice as is done now... 

Additionally, the logic of storing `nextPC_` is minimal, and with the rest of the code around it I do wonder what performance degredation it causes, if any.

I think this may be worth us thinking on this offline for any alternative solutions

Reply:
Offline solution enabled us to get the PC from the fetch unit rather than toring it in the ROB or WB unit at commit. Any flushing, branch mis-predictions, or load violations will implicitly keep the PC value correct. A small change was made inside the fetch unit to keep the PC up to date when the loop buffer is currently supplying instructions.

In effect, no additional logic is required to keep track of the PC to resume at next time the process in question is scheduled.